# *Exploring Hacker News Posts*

In this project, we'll compare two different types of posts from [Hacker News](news.ycombinator.com), a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll specifically compare these two types of posts to determine the following:

Do Ask HN or Show HN receive more comments on average?
Do posts created at a certain time receive more comments on average?
It should be noted that the [data set](https://www.kaggle.com/hacker-news/hacker-news-posts) we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.

### Introduction and Getting Data Ready

Here, we will make use of csv module and convert hackernews.csv file into a lists of lists.

- _**headers**_ list will contain the column names
- _**hn**_ list will cover the whole remaining data

In [1]:
f = open('hacker_news.csv')
from csv import reader
read_file = reader(f)
data = list(read_file)
headers = data[0]
hn = data[1:]
hn[:5]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

In [2]:
headers

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

### Extracting Ask HN and Show HN Posts

Using regex, we will search posts that beign with either Ask HN or Show HN posts.<br>And divide data into three lists:

- show_posts
- ask_posts
- other_posts

In [3]:
import re
patterna = r"^Ask HN"
patterns = r"^Show HN"

ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    match1 = re.search(patterna, title, re.I)
    match2 = re.search(patterns, title, re.I)
    if match1:
        ask_posts.append(row)
    elif match2:
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))
ask_posts[:3]

1744
1162
17194


[['12296411',
  'Ask HN: How to improve my personal website?',
  '',
  '2',
  '6',
  'ahmedbaracat',
  '8/16/2016 9:55'],
 ['10610020',
  'Ask HN: Am I the only one outraged by Twitter shutting down share counts?',
  '',
  '28',
  '29',
  'tkfx',
  '11/22/2015 13:43'],
 ['11610310',
  'Ask HN: Aby recent changes to CSS that broke mobile?',
  '',
  '1',
  '1',
  'polskibus',
  '5/2/2016 10:14']]

### Calculating the Average Number of Comments for Ask HN and Show HN Posts
Now that we separated Ask HN and Show HN posts into different lists, we'll calculate the average number of comments each type of post receives.

In [6]:
ask_comments = [int(i[4]) for i in ask_posts]
show_comments = [int(i[4]) for i in show_posts]

avg_ask_comments = sum(ask_comments)/len(ask_comments)
avg_show_comments = sum(show_comments)/len(show_comments)

print(avg_ask_comments)
avg_show_comments

14.038417431192661


10.31669535283993

On Average, the Ask HN posts received 40% comments more then the Show HN posts.<br>
We will continue our further analysis, using the Ask HN posts only.

### Finding the Amount of Ask HN Posts and Comments by Hour Created
Next, we'll determine if we can maximize the amount of comments an Ask HN post receives by creating it at a certain time. First, we'll find the amount of Ask HN posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments Ask HN posts created at each hour of the day receive.

In [11]:
import datetime as dt

created_at = []
for row in ask_posts:
    created_at.append(row[6])

counts_by_hour = {}
comments_by_hour = {}

result_list = list(zip(ask_comments, created_at))

result_list[:10]

[(6, '8/16/2016 9:55'),
 (29, '11/22/2015 13:43'),
 (1, '5/2/2016 10:14'),
 (3, '8/2/2016 14:20'),
 (17, '10/15/2015 16:38'),
 (1, '9/26/2015 23:23'),
 (4, '4/22/2016 12:24'),
 (1, '11/16/2015 9:22'),
 (1, '2/24/2016 17:57'),
 (2, '6/4/2016 17:17')]

In [35]:
for row in result_list:
    date = dt.datetime.strptime(row[1], '%m/%d/%Y %H:%M')
    hour = date.strftime("%H")
    
    if hour in counts_by_hour:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += row[0]
    else:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = row[0]    
        
comments_by_hour

{'00': 894,
 '01': 1366,
 '02': 2762,
 '03': 842,
 '04': 674,
 '05': 928,
 '06': 794,
 '07': 534,
 '08': 984,
 '09': 509,
 '10': 1587,
 '11': 1282,
 '12': 1378,
 '13': 2535,
 '14': 2835,
 '15': 8954,
 '16': 3645,
 '17': 2295,
 '18': 2878,
 '19': 2376,
 '20': 3444,
 '21': 3490,
 '22': 958,
 '23': 1087}

### Calculating the Average Number of Comments for Ask HN Posts by Hour

In [43]:
avg_by_hour = [[i, comments_by_hour[i]/counts_by_hour[i]] for i in counts_by_hour] 
avg_by_hour

[['04', 7.170212765957447],
 ['09', 5.532608695652174],
 ['19', 10.8],
 ['12', 9.374149659863946],
 ['13', 14.824561403508772],
 ['14', 13.186046511627907],
 ['01', 11.383333333333333],
 ['00', 8.127272727272727],
 ['10', 13.336134453781513],
 ['15', 38.5948275862069],
 ['05', 10.08695652173913],
 ['03', 7.796296296296297],
 ['06', 9.022727272727273],
 ['18', 13.20183486238532],
 ['08', 10.25],
 ['20', 21.525],
 ['17', 11.361386138613861],
 ['07', 7.852941176470588],
 ['23', 7.934306569343065],
 ['16', 16.797235023041473],
 ['22', 6.746478873239437],
 ['11', 11.051724137931034],
 ['02', 23.810344827586206],
 ['21', 16.009174311926607]]

### Sorting and Printing Values from a List of Lists

In [56]:
avg_by_hour = sorted(avg_by_hour, key=lambda x: x[1], reverse=True)
print("Top 5 Hours for 'Ask HN' Comments")

for row in avg_by_hour[:5]:
    time = dt.datetime.strptime(row[0], "%H")
    print("{}: {:.2f} average comments per post".format(time.strftime("%H:%M"), row[1]))

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.<br>

According to the data set [documentation](https://www.kaggle.com/hacker-news/hacker-news-posts/home), the timezone used is Eastern Time in the US.<br>So, we (in UAE) should post at 00:00 AM GST.

### Conclusion
In this project, we analyzed Ask HN and Show HN posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as Ask HN and created between 00:00 and 01:00 GST (3:00 pm EST - 4:00 pm EST).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, Ask HN posts received more comments on average and Ask HN posts created between 00:00 and 01:00 GST received the most comments on average.